### DISCUSSION TITLE
ROMable code and ORDER directive

### DISCUSSION THREAD

**Discussion by jounikor** [open-watcom/open-watcom-v2#810] [Q&A]
I am new to OW, x86, DOS and stuff so bear with me. What I have been toying around is linking ROMable 8086 code that makes use of large memory model and full 20-bit address space. I made few attempts of custom wlink.lnk mainly following the instructions in wikidoc's "lguide". I use dos as the format and option nodefaultlibs. I have tried raw bin but linking always bonkers (seems to assume flat memory model).

My DGROUP group is basically NOEMIT.. How do I make ORDER to copy the initialized '_DATA' segment (of DATA class) into the output binary?  I must miss something fundamental of the 'copy_option'. According to the map file symbols as they should so my startup code could do the needed copying but nothing ends up into the emitted binary. The ROMDATA segment size is 0.

How do I place a specific object file always as the last object of the output binary using the wlink.lnk? Also, is there a way to make wlink not to output fillchars into NOEMIT segments?
EDIT: not outputting fillchars for NOEMIT is doable with OUTPUT directive. However, the calculation of the 'offset' seems to be bugged i.e. when defined it becomes 2's complement of the given value.

I have started to look into the linker code but it takes time before I get my head around it.

**Comment by jmalak** [comment]
Please give here full wlink command because if you use dos as system then it cannot use flat model.
Add here also linker MAP file to see how linker setup addresses.
Also give here full compiler command line to be sure that generated code is correct.

If you do DGROUP group NOEMIT then no data will be emited, because all data segments are part of DGROUP group.
Don't do such thing. Use NOEMIT only with segments not groups.

Generaly note, first you need to deside what output form you will be using, binary or Intel hex format?
Next you must design memory section mapping to use proper linker commands.
Generaly CONST and CONST2 segments contains const data, it can be put into ROM memory.

below is example of linker MAP file for VirtualBox bios build

```
Open Watcom Linker Version 2.0 beta Jan 30 2022 15:42:30 (32-bit)
Copyright (c) 2002-2022 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Created on:       22/01/30 21:16:17
Executable Image: /home/jiri/vbox/out/linux.amd64/release/obj/VBoxPcBios8086/VBoxPcBios8086.rom
creating a DOS executable


                        +------------+
                        |   Groups   |
                        +------------+

Group                           Address              Size
=====                           =======              ====

DGROUP                          f000:0000            000011b4
BIOS32_GROUP                    fdda:0008            00000000



                        +--------------+
                        |   Segments   |
                        +--------------+

Segment                Class          Group          Address         Size
=======                =====          =====          =======         ====

_DATA                  DATA           DGROUP         f000:0000       000000e0
_TEXT                  CODE           AUTO           f000:1600       00009818
BIOS32                 CODE           AUTO           f000:0000da00   000003a6
BIOSSEG                CODE           AUTO           f000:e000       00002000
CONST                  DATA           DGROUP         f00e:0000       00000cda
CONST2                 DATA           DGROUP         f0db:000a       000003fa
BIOS32CONST            FAR_DATA       BIOS32_GROUP   fdda:00000008   00000000
BIOS32CONST2           FAR_DATA       BIOS32_GROUP   fdda:00000008   00000000
BIOS32_DATA            FAR_DATA       BIOS32_GROUP   fdda:00000008   00000000
```

Appropriate linker command is (use binary output)

wlink system dos debug all option quiet, nofarcalls, map, verbose, statics, symfile output raw offset=0xF0000 order clname DATA segaddr=0xF000 segment _DATA clname CODE segment _TEXT   segaddr=0xF000 offset=0x1600 segment BIOS32  segaddr=0xF000 offset=0xDA00 segment BIOSSEG segaddr=0xF000 offset=0xE000    Name VBoxPcBios8086.rom

Memory location is based on segment classes, you probably need on segments level.

**Reply by jounikor** [reply]
Thanks. Any advise appreciated. Just to clarify almost all segment in my DGROUP are NOEMIT, not the class.
This is how my DGROUP group '_NULL,_DATA,_BSS,_BSS2,STACK' looks like.

I'll give a try based on your wlink example and come back shortly. I wonder how output offset worked for you. In my test wlin.lnk offset I get file length ((0-offset) & 0xffffffff) + rest_of_the_stuff. I tweaked the loadraw.c to hotfix this but the actual issue is somewhere else. Im my first test I used:

```
wcl test.c -0 -bcl=test -i=$WATCOM/h -v -s -ml  -zc
```

And the wlink.lnk content for my original test was this:
```
system begin test
    output raw offset=0x20000
    option nodefaultlibs,stack=0,fillchar=0xaa,map,symfile                                                                   
    order   clname BEGDATA NOEMIT segaddr=0x00 segment _NULL
            clname DATA NOEMIT segment _DATA
            clname BSS
            clname STACK segaddr=0xe0
            clname BEGCODE segaddr=0x2000
            clname CODE segment _TEXT
            clname ROMDATA COPY DATA
            clname ROMDATAE
    libpath %WATCOM%/lib286/foobar
    libfile foobar.o
    format dos
end
```
And that produces the following.. I could not figure out how to move CONST and CONST2 out of the to be initialized DATA area. Otherwise, everything is more or less how they should be (excluding that NOEMIT thing).
```
                        +------------+
                        |   Groups   |
                        +------------+

Group                           Address              Size
=====                           =======              ====

DGROUP                          0000:0000            00000e00



                        +--------------+
                        |   Segments   |
                        +--------------+

Segment                Class          Group          Address         Size
=======                =====          =====          =======         ====

_NULL                  BEGDATA        DGROUP         0000:0000       00000100
_DATA                  DATA           DGROUP         0010:0000       0000000c
CONST                  DATA           DGROUP         0010:000c       00000000
CONST2                 DATA           DGROUP         0010:000c       00000000
_BSS                   BSS            DGROUP         0010:000c       0000000a
_BSS2                  BSS            DGROUP         0011:0006       00000000
STACK                  STACK          DGROUP         00e0:0000       00000000
STARTUP                BEGCODE        AUTO           2000:0000       0000001b
_TEXT                  CODE           AUTO           2000:001c       00000000
foobar_TEXT            CODE           AUTO           2000:001c       00000000
ROMDATA                CODE           AUTO           2000:001c       00000000
ROMDATAE               CODE           AUTO           2000:001c       00000000
_EPILOGUE              CODE           AUTO           2000:001c       00000000
test_TEXT              CODE           AUTO           2000:001c       000000a9
FAR_DATA               FAR_DATA       AUTO           200c:0006       00000000
test13_DATA            FAR_DATA       AUTO           200d:0000       0000fd0c
test14_DATA            FAR_DATA       AUTO           2fde:0000       00009c40
```
EDIT: adding the symbol output, which shows that my stack, bss and initializable data are in segment address 0x0000 as they should. The symbol '_yes' in _DATA segment is the one that should end up to ROMable area and copied by the startup code into the RAM.
```
0000:010c*     __begbss
0000:010c*     __enddata
0000:0116*     __endbss
0000:0000*     __begnull
0000:0100*     __begdata
2000:0000      _cstart_
2000:0000      _big_code_
0000:0000*     _test_epilogue_
Module: test.o(/Users/jouni/devel/open-watcom-v2/tmp/test.c)
2000:001c      main_
0000:0100+     _yes
0000:010c*     _muuttuja6
0000:010e*     _muuttuja4
0000:0110*     _muuttuja5
0000:0112*     _muuttuja2
0000:0114*     _muuttuja1
200d:0000+     _jes
200d:000c+     _foobar
200d:7d0c+     _barbuz
2fde:0000+     _foofoo
```

**Reply by jmalak** [reply]
I think you complicate your live by hacking what is standard functionality.
You don't need any such hacking.
If you are working on linker configuration then don't use wcl utility use wlink directly.
Please do things standard way and you will have all under simple transparent control.
Now you have lot of useless dependencies.

You should use OW tool transparent way, do following:
- use standard OW configuration for your host (probably Linux)
- change configuration for your target (DOS), it means setup of environment variable INCLUDE=$WATCOM/h
- use standard wlink.lnk, don't change it any way until you understand what it cause
- compile C sources by wcc don't use wcl, compile you source by simplified command 
  `wcc -bt=dos -s -ml -zc test.c -fo=test.obj`
  need not recompile it if you test linker configuration, only if you change C sources
- prepare test.lnk file which contains your wlink directives and options related to special needs for test image
  because it is absolutely unclear what segment should be included and where and which excluded it is only sample, but show possible syntax or ORDER directive
```
    output raw
    option nodefaultlibs, stack=0, fillchar=0xaa
    library { list of additional libraries }
    order   
        clname BEGDATA NOEMIT
        clname DATA segaddr=0
             segment _DATA NOEMIT
        clname STACK segaddr=0xe0
        clname BEGCODE segaddr=0x2000
        clname CODE 
             segment _TEXT 
             segment ROMDATA
             segment ROMDATAE
```

- link image by command `wlink sys dos name test debug all option map, symfile file test.obj @test.lnk`

**Reply by jmalak** [reply]
Anyway I can recommend following approach.
ORDER is map of output address space that start on minimal output address.
if you add class/segments it is for output order.
If you include class take into account that automatically include all segments in this class and in its original order.
if you need exclude some segment in the class add them with NOEMIT.
if you specify segment then you specify order of such segment in class output address space
if segment is not specified in ORDER CLASS then it is add on the end of class output address space in its original order

**Reply by jounikor** [reply]
Thanks again!  Very much appreciated. I Did what you suggested as steps & test.lnk. I had to add 'clname BSS' otherwise bss ended into a middle of CODE class segments causing the DGROUP being over 64K and linking to fail: 
`wlink name test debug all option map, symfile file foobar.o,test.o @test.lnk`
foobar.o is the "cstartup" code, which basically just calls "mail_" at the moment (placed in class BEGCODE and segment STARTUP). After linking output binary does no really differ what I managed with my complicated/integrated approach.
Going to dig a bit more into this..

**Reply by jmalak** [reply]
try to use *.lnk file per project with exact definition order of class/segments.
you can try define your own target but take into account that header files are control by -bt compiler option.

you can create your own linker target which define automaticaly your startup module and your run-lime library etc. which are common setup for all your ROM projects. May be right better is use `format raw bin` in your linker file instead of `system dos`.
`format raw bin` doesn't handle something like DOS header, relocation table etc. use simply flat address space from 0 which can be segmented.
by example if you define `system myrom` then you can add following section to wlink.lnk

```
system begin myrom
    option nodefaultlibs
    option stack=0
    option fillchar=0xaa 
    libfile <name of your startup module>
    libpath <path where you have your own run-time library>
    library <your standard runtime libraries>
    format raw bin
end
```

you can add also ORDER directive if segment mapping will be always the same if not then it has no sense to add it to this system section. Because all items in system section is always add to wlink command.

then wlink command will use `wlink system myrom ......`
in case of use wcl then you must use two options one for compilers target and second one for linker target(system).
then you can use `wcl -bt=dos -l=myrom ......`

**Reply by jounikor** [reply]
I got ```clname ROMDATA copy DATA``` kind of working but hit another bug (I guess).. The following:
```
output raw offset=0x10000
option nodefaultlibs
order
    clname BEGDATA segaddr=0x0000 NOEMIT segment _NULL
    clname DATA segment _DATA   <--- note no NOEMIT here
    ...
    clname BEGCODE segaddr=0x1000 segment STARTUP
    clname CODE segment _CODE
    clname ROMDATA copy DATA
    clname ROMDATAE
libpath %WATCOM%/foobar
libfile foobar.o
format dos
```
The `offset=0x10000` skips the entire DGROUP and more from beginning of the ROM so the initializable data in there goes into the void. However, now for some reason the `clname ROMDATA copy DATA` kicks in actually copies the DATA class at the end of the ROM file i.e. after all FAR_DATAs.. ehh. Both ROMDATA and ROMDATAE still remain empty. There is no sign of this copying in the map file.

For now, I'll spend a bit more time studying the `wl` code and stop just trying stuff ;) I am a bit by bit getting some grasp of the code.

**Reply by jmalak** [reply]
I think you use offset incorrectly, it is offset in output space where binary image starts by example if you want to create binary image for bios starting 0x8000:0 (and length of one segment) you map classes by order directive to 0x8000:0 and above and using offset 0x80000 you cut of address space 0-0x80000 that binary image contains only part starting 0x80000. for you it has no sense if you start at 0 then offset must be also 0. Offset control what will be emitted but image use internaly addresses 0x8000:..... and only fragment of address space is emited.
if you want start at 0x10000 then you must start map on 0x1000:0 then offset 0x10000 remove space 0-0x10000 from generated image even if address will use 0x1000:.... in image. without using offset full output space will be emited 0-..... end of image, on the begining will be space 0-0x10000 fill in by fill character. Try to check.
